### PR TITLE
Repair the move up/down links in OrderedTabularInline

### PR DESCRIPTION
**Pull request by flupzor**
This fixes this issue:

https://github.com/bfirsh/django-ordered-model/issues/102

### CODE REVIEW

**Comment by saily** [comment]
@bfirsh can you please merge this one and release a new version?

**Comment by geraneum** [comment]
Got NoReverseMatch error after applying this method...

**Comment by flupzor** [comment]
@saily Could you be more specific? Afaik I did not touch the templates.

@geraneum Where did you encounter  this? How did you set it up?

**Comment by geraneum** [comment]
Sub-Model:
```
class QuestionOption(OrderedModel):
    question = models.ForeignKey('Question', related_name='options')
    title = models.CharField(max_length=128)
    order_with_respect_to = 'question'

    class Meta(OrderedModel.Meta):
        pass
```

Parent Model:
```
class Question(MPTTModel, BaseModel):
    title = models.CharField(_('title'), max_length=128)
```

Sortable Admin (I did an override for `move_up_down_links`):
```
class QuestionOptionInlineAdmin(OrderedTabularInline):
    model = models.QuestionOption
    fields = ('title', 'order', 'move_up_down_links')
    readonly_fields = ('order', 'move_up_down_links')
    ordering = ('order',)
    extra = 0
```

The error I get:
```
NoReverseMatch at /admin/listings/question/fa0a7476-03d5-4e3d-98b6-00f61d9ae1fe/change/

Reverse for 'listings_questionoption_order_up_inline' with arguments '('fa0a7476-03d5-4e3d-98b6-00f61d9ae1fe', 1, 'up')' and keyword arguments '{}' not found. 0 pattern(s) tried: []

Request Method:	GET
Request URL:	http://0.0.0.0:8000/admin/listings/question/fa0a7476-03d5-4e3d-98b6-00f61d9ae1fe/change/
Django Version:	1.10.5
Exception Type:	NoReverseMatch
Exception Value:	
Reverse for 'listings_questionoption_order_up_inline' with arguments '('fa0a7476-03d5-4e3d-98b6-00f61d9ae1fe', 1, 'up')' and keyword arguments '{}' not found. 0 pattern(s) tried: []
Exception Location:	/Users/Mazdak/.virtualenvs/unified-service/lib/python3.5/site-packages/django/urls/resolvers.py in _reverse_with_prefix, line 392
Python Executable:	/Users/Mazdak/.virtualenvs/unified-service/bin/python
Python Version:	3.5.2
```

**Comment by bfirsh** [comment]
Thanks both! Perhaps we could construct a test case based on this example.

**Comment by saily** [comment]
@flupzor sorry, i was wrong, i have the same `NoReverseMatch` error.

**Comment by flupzor** [comment]
How does your QuestionAdmin look like? Also how foes your override
look like?




On Sun, Feb 19, 2017, at 12:02 PM, Mazdak B. wrote:

> Sub-Model:


> class QuestionOption(OrderedModel): question =
> models.ForeignKey('Question', related_name='options') title =
> models.CharField(max_length=128) order_with_respect_to = 'question'
> class Meta(OrderedModel.Meta): pass
>
> Parent Model:


> class Question(MPTTModel, BaseModel): title =
> models.CharField(_('title'), max_length=128)
>
> Sortable Admin (I did an override for move_up_down_links):


> class QuestionOptionInlineAdmin(OrderedTabularInline): model =
> models.QuestionOption fields = ('title', 'order',
> 'move_up_down_links') readonly_fields = ('order',
> 'move_up_down_links') ordering = ('order',) extra = 0
>
> The error I get:


> NoReverseMatch at /admin/listings/question/fa0a7476-03d5-4e3d-98b6-
> 00f61d9ae1fe/change/  Reverse for
> 'listings_questionoption_order_up_inline' with arguments '('fa0a7476-03d5-4e3d-98b6-
> 00f61d9ae1fe', 1, 'up')' and keyword arguments '{}' not found. 0
> pattern(s) tried: []  Request Method: GET Request URL:
> http://0.0.0.0:8000/admin/listings/question/fa0a7476-03d5-4e3d-98b6-00f61d9ae1fe/change/
> Django Version: 1.10.5 Exception Type: NoReverseMatch Exception Value:
> Reverse for 'listings_questionoption_order_up_inline' with arguments
> '('fa0a7476-03d5-4e3d-98b6-00f61d9ae1fe', 1, 'up')' and keyword
> arguments '{}' not found. 0 pattern(s) tried: [] Exception Location:
> /Users/Mazdak/.virtualenvs/unified-service/lib/python3.5/site-
> packages/django/urls/resolvers.py in _reverse_with_prefix, line 392
> Python Executable: /Users/Mazdak/.virtualenvs/unified-
> service/bin/python Python Version: 3.5.2
>
> — You are receiving this because you were mentioned. Reply to this
> email directly, view it on GitHub[1], or mute the thread[2].
> 


Links:

  1. https://github.com/bfirsh/django-ordered-model/pull/110#issuecomment-280911610

**Comment by saily** [comment]
@flupzor i assume @geraneum is missing following block in his `QuestionAdmin` class which was mentioned by @bfirsh in the docs here https://github.com/bfirsh/django-ordered-model#admin-integration. 

```python


    def get_urls(self):
        urls = super().get_urls()
        for inline in self.inlines:
            if hasattr(inline, 'get_urls'):
                urls = inline.get_urls(self) + urls
        return urls
```

after i've added this to mine, issue was gone. Thanks a lot for pointing me into right direction. We should add this to docs as an FAQ item.

**Comment by geraneum** [comment]
@saily Thanks... including `get_urls` in parent model admin together with overriding `move_up_down_links` solved the problem.

**Comment by flupzor** [comment]
@bfirsh What is needed for this PR to be merged?

**Comment by szechyjs** [comment]
Any chance of getting this merged and released?

**Comment by bfirsh** [comment]
The thing that was blocking me is this is a big bundle of code without test. From glancing at it it isn't clear what it's doing.

Since there are several confirmations this is working however, I will get this released later today. :)

File: ordered_model/admin.py [hunk 1]
```diff
@@ -248,14 +248,23 @@ def get_preserved_filters(cls, request):
...
 
     def move_up_down_links(self, obj):
-        if obj.id:
-            order_obj_name = 'obj'
-            if obj._get_order_with_respect_to() is not None:
-                order_obj_name = obj._get_order_with_respect_to().id
+        if not obj.id:
+            return ''
+
+        # Find the fields which refer to the parent model of this inline, and
```
Comment (bfirsh) [inline ordered_model/admin.py]: Happy to merge this as is, but this bit of logic probably makes sense to be factored into its own function somewhere. It's hard to parse what is going on here.

**Comment by saily** [comment]
@bfirsh i'll try to write some tests today and dig into the logic of the original author.

**Comment by bfirsh** [comment]
@saily Sweet, thanks! No worries if you don't get to it. If you don't, I'll merge and release later today/tomorrow and we can fix up later. :)

**Comment by flupzor** [comment]
> The thing that was blocking me is this is a big bundle of code without test. From glancing at it it isn't clear what it's doing.

@bfirsh Nice of you to mention 6 months after I created the  pull request.

**Comment by bfirsh** [comment]
@flupzor We're all doing this in our spare time, and we're all busy people. There is no need for that attitude.

**Comment by flupzor** [comment]
@bfirsh My point is, If you mentioned this the first time you reacted, I would've been motivated to add tests and clarify the code, now, I'm not.

**Comment by bfirsh** [comment]
I did say:

> Thanks both! Perhaps we could construct a test case based on this example.

Sorry if that wasn't clear enough. I do appreciate your contribution, and understand that it's a bit frustrating. When I first commented on this PR I was on a career break, so I didn't really have time to review it properly. I just wanted to keep things moving for when I got back.

I will wait until the end of the day to see if @saily has had a chance to add a test. If not, I'll get this merged and released anyway. :)

**Comment by saily** [comment]
@bfirsh i'd prefer you merge it right now, i'll take care of adding a test after easter weekend.

**Comment by bfirsh** [comment]
@saily Sweet - sure thing.

Thanks all for your help!

**Comment by iMerica** [comment]
This is still broken for me in 1.4.1 using Django 1.9.7 with a basic model, no subsets.
